File: benchmarks/fannkuch_redux/ruby/bench.rb
# The Computer Language Benchmarks Game
# http://shootout.alioth.debian.org/
# Contributed by Wesley Moxam

@EXPECT_CKSUM = 1616
@MAX_N = 8

def run_iter(n)
  for i in 1..n do
    inner_iter()
  end
end

def inner_iter()
  n = @MAX_N
  sign, maxflips, sum = 1, 0, 0

  p = [nil].concat((1..n).to_a)
  q = p.dup
  s = p.dup

  while(true)
    # Copy and flip.
    q1 = p[1]				# Cache 1st element.
    if q1 != 1
      q = p.dup
      flips = 1
      while(true)
	      qq = q[q1]
	      if qq == 1				# ... until 1st element is 1.
	        sum = sum + sign * flips
	        maxflips = flips if flips > maxflips # New maximum?
	        break
	      end
	      q[q1] = q1
	      if q1 >= 4
	        i, j = 2, q1 - 1
	        begin
            q[i], q[j] = q[j], q[i]
            i = i + 1
            j = j - 1
          end while i < j
	      end
	      q1 = qq
        flips = flips + 1
      end
    end
    # Permute.
    if sign == 1
      # Rotate 1<-2.
      p[1], p[2] = p[2], p[1]
      sign = -1	
    else
      # Rotate 1<-2 and 1<-2<-3.
      p[2], p[3] = p[3], p[2]
      sign = 1
      3.upto(n) do |i|
        (s[i] =  s[i] - 1) && break unless s[i] == 1
              if i == n then
	        if sum != @EXPECT_CKSUM then
                  puts("bad checksum: %d vs %d" % [sum, @EXPECT_CKSUM])
                  exit 1
                end
	        return [sum, maxflips] # Out of permutations.
              end
	      s[i] = i
        # Rotate 1<-...<-i+1.
	      t = p[1]
        1.upto(i) do |j|
          p[j] = p[j+1]
        end
        p[i+1] = t
      end
    end
  end
end

File: benchmarks/spectralnorm/javascript/bench.js
// The Computer Language Benchmarks Game
// http://benchmarksgame.alioth.debian.org/
//
// contributed by Ian Osgood
// Optimized by Roy Williams

var SPECTRAL_N  = 1000;
var EXPECT_CKSUM = 1.2742241481294835914184204739285632967948913574218750;

function A(i,j) {
  return 1/(((i+j)*(i+j+1)>>>1)+i+1);
}

function Au(u,v) {
    var n = u.length;
  for (var i=0; i<n; ++i) {
    var t = 0;
    for (var j=0; j<n; ++j)
      t += A(i,j) * u[j];
    v[i] = t;
  }
}

function Atu(u,v) {
  var n = u.length;
  for (var i=0; i<n; ++i) {
    var t = 0;
    for (var j=0; j<n; ++j)
      t += A(j,i) * u[j];
    v[i] = t;
  }
}

function AtAu(u,v,w) {
  Au(u,w);
  Atu(w,v);
}

function run_iter(n) {
  var i = 0;
  for (i = 0; i < n; i++) {
    var checksum = inner_iter(SPECTRAL_N);
    if (checksum != EXPECT_CKSUM) {
      print("bad checksum: " + checksum + " vs " + EXPECT_CKSUM);
    }
  }
}

function inner_iter(n) {
  var storage_ = new ArrayBuffer(n * 24);
  var u = new Float64Array(storage_, 0, n),
      v = new Float64Array(storage_, 8*n, n),
      w = new Float64Array(storage_, 16*n, n);
  var i, vv=0, vBv=0;
  for (i=0; i<n; ++i) {
    u[i] = 1; v[i] = w[i] = 0; 
  }
  for (i=0; i<10; ++i) {
    AtAu(u,v,w);
    AtAu(v,u,w);
  }
  for (i=0; i<n; ++i) {
    vBv += u[i]*v[i];
    vv  += v[i]*v[i];
  }
  return Math.sqrt(vBv/vv);
}

//print(spectralnorm(arguments[0]).toFixed(9));
